### PR TITLE
Core: Support webpack5 and webpack4 side by side

### PR DESCRIPTION
**Pull request by shilman**
Issue: #9216

## What I did

With @ndelangen:

- [x] Upgrade to webpack5 + dependencies
- [x] Fixed webpack build in official example
- [x] Split up `@storybook/core` into `core-server`, `core-client`, `core-common`
- [x] Introduce a builder abstraction and implement `builder-webpack4`, `builder-webpack5`
- [x] Set up snapshot tests for the webpack configuration so that we can test against `next`
- [ ] Set up an e2e test that runs outside the monorepo to avoid hoisting issues

## How to test

See `core-presets.test.ts` whose snapshots we've manually checked against `next` (webpack4)

### CODE REVIEW

**Comment by phated** [comment]
It looks like this needs a rebase on next so it can remove https://github.com/storybookjs/storybook/pull/13808/commits/c5c5a983ab18885ad586b56bbb57e8fb1a1c88e6 which squashed the entire history between then & now into 1 commit and resulted in this massive diff.

**Comment by yash558** [comment]
Hello sir I want to be  contribute with your organization in a open source project .How could  I started pls can you help me sir.

File: lib/builder-webpack5/src/preview/base-webpack.config.ts [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+      }),
+    ],
+  }),
+  dedent`
+    Default PostCSS plugins are deprecated. When switching to '@storybook/addon-postcss',
+    you will need to add your own plugins, such as 'postcss-flexbugs-fixes' and 'autoprefixer'.
+
+    See https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#deprecated-default-postcss-plugins for details.
+  `
+);
```
Comment (phated) [inline lib/builder-webpack5/src/preview/base-webpack.config.ts:27]: Webpack 5 probably shouldn't even ship with these deprecated implicit loaders. Thoughts?

**Review by phated — COMMENTED** [review]
Whoa 🤯

File: lib/builder-webpack5/src/preview/base-webpack.config.ts (hunk 1 above)
Comment (ndelangen) [reply lib/builder-webpack5/src/preview/base-webpack.config.ts:27]: yeah good point!

File: app/angular/package.json [hunk 2]
```diff
@@ -43,11 +43,12 @@
   "dependencies": {
     "@storybook/addons": "6.2.0-alpha.24",
     "@storybook/core": "6.2.0-alpha.24",
+    "@storybook/core-common": "6.2.0-alpha.24",
     "@storybook/node-logger": "6.2.0-alpha.24",
     "@types/webpack-env": "^1.16.0",
     "autoprefixer": "^9.8.6",
     "core-js": "^3.8.2",
-    "fork-ts-checker-webpack-plugin": "^4.1.6",
+    "fork-ts-checker-webpack-plugin": "^6.0.4",
```
[the remark this answers is not in the document]
Comment (ndelangen) [reply app/angular/package.json:52]:
```suggestion
    "fork-ts-checker-webpack-plugin": "^4.1.6",
```

File: examples/vue-3-cli/package.json [hunk 3]
```diff
@@ -9,23 +9,21 @@
...
-    "core-js": "^3.6.5",
+    "core-js": "^3.8.2",
     "vue": "^3.0.0"
   },
   "devDependencies": {
     "@babel/core": "^7.12.10",
     "@storybook/addon-actions": "6.2.0-alpha.24",
     "@storybook/addon-essentials": "6.2.0-alpha.24",
     "@storybook/addon-links": "6.2.0-alpha.24",
-    "@storybook/addon-storyshots": "6.2.0-alpha.24",
```
[the remark this answers is not in the document]
Comment (ndelangen) [reply examples/vue-3-cli/package.json:20]:
```suggestion
    "@storybook/addon-storyshots": "6.2.0-alpha.24",
```

File: examples/vue-3-cli/package.json [hunk 4]
```diff
@@ -9,23 +9,21 @@
...
     "@storybook/vue3": "6.2.0-alpha.24",
     "@vue/cli-plugin-babel": "~4.5.0",
     "@vue/cli-plugin-typescript": "~4.5.0",
     "@vue/cli-service": "~4.5.0",
     "@vue/compiler-sfc": "^3.0.0",
     "babel-loader": "^8.2.2",
-    "typescript": "~3.9.3",
-    "vue-jest": "^5.0.0-alpha.8",
```
[the remark this answers is not in the document]
Comment (ndelangen) [reply examples/vue-3-cli/package.json:28]:
```suggestion
    "typescript": "~3.9.3",
    "vue-jest": "^5.0.0-alpha.8",
```

File: addons/docs/src/frameworks/common/preset.ts [hunk 5]
```diff
@@ -6,7 +6,7 @@ import remarkExternalLinks from 'remark-external-links';
 import createCompiler from '../../mdx/mdx-compiler-plugin';
 
 const resolvedBabelLoader = require.resolve('babel-loader', {
-  paths: [require.resolve('@storybook/core')],
+  paths: [require.resolve('@storybook/builder-webpack4')], // FIXME!!!
```
Comment (ndelangen) [inline addons/docs/src/frameworks/common/preset.ts:9]: TODO

File: addons/storyshots/storyshots-core/package.json [hunk 6]
```diff
@@ -64,23 +65,22 @@
...
+    "@storybook/vue": "6.2.0-alpha.28",
     "@storybook/vue3": "6.2.0-alpha.28",
     "babel-loader": "^8.2.2",
     "enzyme": "^3.11.0",
     "enzyme-to-json": "^3.6.1",
     "jest-emotion": "^10.0.32",
     "jest-preset-angular": "^8.3.2",
     "jest-vue-preprocessor": "^1.7.1",
-    "rxjs": "^6.6.3",
-    "vue-jest": "^5.0.0-alpha.8"
```
Comment (ndelangen) [inline addons/storyshots/storyshots-core/package.json:75]: REVERT

File: addons/storyshots/storyshots-core/package.json [hunk 7]
```diff
@@ -64,23 +65,22 @@
...
+    "rxjs": "^6.6.3"
   },
   "peerDependencies": {
     "@angular/core": ">=6.0.0",
     "@angular/platform-browser-dynamic": ">=6.0.0",
     "@storybook/angular": "*",
     "@storybook/react": "*",
     "@storybook/vue": "*",
-    "@storybook/vue3": "*",
```
Comment (ndelangen) [inline addons/storyshots/storyshots-core/package.json:83]: REVERT

File: addons/storyshots/storyshots-core/package.json [hunk 8]
```diff
@@ -103,9 +103,6 @@
     "@storybook/vue": {
       "optional": true
     },
-    "@storybook/vue3": {
-      "optional": true
-    },
```
Comment (ndelangen) [inline addons/storyshots/storyshots-core/package.json:108]: REVERT

File: addons/storyshots/storyshots-core/package.json [hunk 9]
```diff
@@ -123,9 +120,6 @@
     },
     "vue": {
       "optional": true
-    },
-    "vue-jest": {
-      "optional": true
```
Comment (ndelangen) [inline addons/storyshots/storyshots-core/package.json:128]: REVERT

File: app/angular/package.json [hunk 10]
```diff
@@ -42,8 +42,9 @@
   },
   "dependencies": {
     "@storybook/addons": "6.2.0-alpha.28",
-    "@storybook/api": "6.2.0-alpha.28",
```
Comment (ndelangen) [inline app/angular/package.json:45]: REVERT

File: app/angular/src/server/__tests__/create-fork-ts-checker-plugin.test.ts [hunk 11]
```diff
@@ -1,35 +1,41 @@
...
-
-    expect(instance).toBeInstanceOf(ForkTsCheckerWebpackPlugin);
-    expect(instance.tsconfig).toEqual(tsLoaderOptions.configFile);
-  });
-
-  it('should create a ForkTsCheckerWebpackPlugin instance without passing options', () => {
-    // add proper typing
-    const instance = createForkTsCheckerInstance({} as any);
-    expect(instance).toBeInstanceOf(ForkTsCheckerWebpackPlugin);
-  });
```
Comment (ndelangen) [inline app/angular/src/server/__tests__/create-fork-ts-checker-plugin.test.ts:34]: REVERT

File: app/angular/src/server/angular-cli_utils.ts [hunk 12]
```diff
@@ -67,6 +67,7 @@ function isStylingRule(rule: RuleSetRule) {
 }
 
 export function filterOutStylingRules(config: Configuration) {
+  // @ts-ignore
```
Comment (ndelangen) [inline app/angular/src/server/angular-cli_utils.ts:70]: REVERT?

File: app/aurelia/src/server/__tests__/create-fork-ts-checker-plugin.test.ts [hunk 13]
```diff
@@ -1,35 +1,39 @@
...
+//   it('should create a ForkTsCheckerWebpackPlugin instance without passing options', () => {
+//     // add proper typing
+//     const instance = createForkTsCheckerInstance({} as any);
+//     expect(instance).toBeInstanceOf(ForkTsCheckerWebpackPlugin);
+//   });
+// });
+
+it('work-around', () => {
+  expect(true).toBe(true);
 });
```
Comment (ndelangen) [inline app/aurelia/src/server/__tests__/create-fork-ts-checker-plugin.test.ts:39]: REVERT

File: lib/builder-webpack4/README.md [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+- `url.js` - controls the URL in the preview and sets the selection based on it.
+- `loadCsf` - loads CSF files from `require.context()` calls and uses `ClientApi` to load them into the store.
+
+Each module uses the channel to communicate with each other and the manager. Each module also has direct access to the story store.
+
+### Events on startup
+
+The store can only be changed during "configuration". The `ConfigApi` will call `store.startConfiguration()`, then the user code (or `loadCsf`'s loader) which will use client API to load up stories. At the end of the user's code the `ConfigApi` will call `store.finishConfiguration()`. At this point the `SET_STORIES` event is emitted and the stories are transmitted to the manager.
+
+The URL of the preview is a "selection specifier" that controls the initial selection, which is chosen when configuration ends. Also (outside of configuration) the `SET_CURRENT_STORY` "command" event can be used to set the selection on the store. When the selection is set, a `CURRENT_STORY_WAS_SET` event is emitted which triggers a rendering.
```
Comment (ndelangen) [inline lib/builder-webpack4/README.md:28]: TODO: this is a copy of lib/core, should be a proper readme

File: lib/builder-webpack4/src/preview/iframe-webpack.config.ts [hunk 15]
```diff
@@ -57,20 +63,19 @@ export default async ({
...
+
+  const headHtmlSnippet = await presets.apply('previewHeadTemplate');
+  const bodyHtmlSnippet = await presets.apply('previewBodyTemplate');
+  const template = await presets.apply<string>('previewMainTemplate');
+
+  // TODO: envs should come rom presets
   const { raw, stringified } = loadEnv({ production: true });
   const babelLoader = createBabelLoader(babelOptions, framework);
   const isProd = configType === 'PRODUCTION';
+  // TODO FIX ME - does this need to be ESM?
```
Comment (ndelangen) [inline lib/builder-webpack4/src/preview/iframe-webpack.config.ts:78]: This is good the way it is now.

File: lib/builder-webpack4/src/preview/iframe-webpack.config.ts [hunk 16]
```diff
@@ -57,20 +63,19 @@ export default async ({
...
-    getPreviewHeadHtml(configDir, process.env)
-  );
-  const bodyHtmlSnippet = await presets.apply(
-    'previewBody',
-    getPreviewBodyHtml(configDir, process.env)
-  );
+
+  const headHtmlSnippet = await presets.apply('previewHeadTemplate');
+  const bodyHtmlSnippet = await presets.apply('previewBodyTemplate');
+  const template = await presets.apply<string>('previewMainTemplate');
```
Comment (ndelangen) [inline lib/builder-webpack4/src/preview/iframe-webpack.config.ts:72]: We should document these options can now be set via presets!

File: lib/builder-webpack4/src/preview/iframe-webpack.config.ts [hunk 17]
```diff
@@ -164,14 +174,14 @@ export default async ({
       isProd ? null : new WatchMissingNodeModulesPlugin(nodeModulesPaths),
       isProd ? null : new HotModuleReplacementPlugin(),
       new CaseSensitivePathsPlugin(),
-      quiet ? null : new ProgressPlugin(),
+      quiet ? null : new ProgressPlugin({}),
```
Comment (ndelangen) [inline lib/builder-webpack4/src/preview/iframe-webpack.config.ts:177]: What sort of options can / should we set here?

File: lib/builder-webpack5/README.md [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+- `url.js` - controls the URL in the preview and sets the selection based on it.
+- `loadCsf` - loads CSF files from `require.context()` calls and uses `ClientApi` to load them into the store.
+
+Each module uses the channel to communicate with each other and the manager. Each module also has direct access to the story store.
+
+### Events on startup
+
+The store can only be changed during "configuration". The `ConfigApi` will call `store.startConfiguration()`, then the user code (or `loadCsf`'s loader) which will use client API to load up stories. At the end of the user's code the `ConfigApi` will call `store.finishConfiguration()`. At this point the `SET_STORIES` event is emitted and the stories are transmitted to the manager.
+
+The URL of the preview is a "selection specifier" that controls the initial selection, which is chosen when configuration ends. Also (outside of configuration) the `SET_CURRENT_STORY` "command" event can be used to set the selection on the store. When the selection is set, a `CURRENT_STORY_WAS_SET` event is emitted which triggers a rendering.
```
Comment (ndelangen) [inline lib/builder-webpack5/README.md:28]: Should be a proper README for this package

File: lib/builder-webpack5/src/index.ts [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+        logger.error('=> Failed to build the preview');
+        process.exitCode = 1;
+
+        if (error) {
+          logger.error(error.message);
+          return fail(error);
+        }
+
+        if (stats && (stats.hasErrors() || stats.hasWarnings())) {
+          const { warnings, errors } = stats.toJson(statsOptions);
```
Comment (ndelangen) [inline lib/builder-webpack5/src/index.ts:124]: need a guard against invalid statsOptions (without warnings or errors) see manager

File: lib/builder-webpack5/src/index.ts [hunk 20]
```diff
@@ -0,0 +1,143 @@
...
+
+          errors.forEach((e) => logger.error(e.message));
+          warnings.forEach((e) => logger.error(e.message));
+          return fail(stats);
+        }
+      }
+
+      logger.trace({ message: '=> Preview built', time: process.hrtime(startTime) });
+      if (stats) {
+        stats.toJson(statsOptions).warnings.forEach((e) => logger.warn(e.message));
```
Comment (ndelangen) [inline lib/builder-webpack5/src/index.ts:134]: need a guard against invalid statsOptions (without warnings or errors) see manager

File: lib/builder-webpack5/src/preview/base-webpack.config.ts [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+          test: /\.(mp4|webm|wav|mp3|m4a|aac|oga)(\?.*)?$/,
+          loader: require.resolve('url-loader'),
+          options: {
+            limit: 10000,
+            name: 'static/media/[name].[hash:8].[ext]',
+          },
+        },
+      ],
+    },
+  };
```
Comment (ndelangen) [inline lib/builder-webpack5/src/preview/base-webpack.config.ts:69]: prefer mutation over re-creating new object here.

File: lib/core-client/README.md [hunk 22]
```diff
@@ -0,0 +1,28 @@
...
+- `url.js` - controls the URL in the preview and sets the selection based on it.
+- `loadCsf` - loads CSF files from `require.context()` calls and uses `ClientApi` to load them into the store.
+
+Each module uses the channel to communicate with each other and the manager. Each module also has direct access to the story store.
+
+### Events on startup
+
+The store can only be changed during "configuration". The `ConfigApi` will call `store.startConfiguration()`, then the user code (or `loadCsf`'s loader) which will use client API to load up stories. At the end of the user's code the `ConfigApi` will call `store.finishConfiguration()`. At this point the `SET_STORIES` event is emitted and the stories are transmitted to the manager.
+
+The URL of the preview is a "selection specifier" that controls the initial selection, which is chosen when configuration ends. Also (outside of configuration) the `SET_CURRENT_STORY` "command" event can be used to set the selection on the store. When the selection is set, a `CURRENT_STORY_WAS_SET` event is emitted which triggers a rendering.
```
Comment (ndelangen) [inline lib/core-client/README.md:28]: should get it's own proper README

File: lib/core-client/docs/standalone.md [hunk 23]
```diff
@@ -0,0 +1,92 @@
...
+```js
+const storybook = require('@storybook/angular/standalone');
+
+storybook({
+  mode: 'dev',
+  port: 9009,
+  configDir: './.storybook',
+  staticDir: ['localDir', 'localDir/images:/images/public']
+});
+```
```
Comment (ndelangen) [inline lib/core-client/docs/standalone.md:92]: duplicate of core / core-server, can be REMOVED

File: lib/core-client/docs/storiesOf.md [hunk 24]
```diff
@@ -0,0 +1,89 @@
...
+
+## Component Story Format migration
+
+To make it easier to adopt the new [Component Story Format (CSF)](https://storybook.js.org/docs/react/api/csf), we've created an automatic migration tool to transform `storiesOf` API to Module format.
+
+```sh
+sb migrate storiesof-to-csf --glob=src/**/*.stories.js
+```
+
+For more information, see the CLI's [Codemod README](https://github.com/storybookjs/storybook/tree/next/lib/codemod).
```
Comment (ndelangen) [inline lib/core-client/docs/storiesOf.md:89]: duplicate can be REMOVED

**Comment by kirill-konshin** [comment]
I see a lot of progress during last weeks, awesome! Do you have any expectation, when this will be available as alpha?

**Comment by shilman** [comment]
@kirill-konshin trying to get it wrapped up this week: adding tests, wrapping up loose ends, reviewing.

**Review by tmeasday — APPROVED** [review]
Did a high level review of the tests and the architecture with @shilman and it LGTM.

**Comment by shilman** [comment]
Early, completely non-optimized perf testing.

## Install size up, build size down

This is a CRA benchmark (webpack4) and we are building the Storybook "manager" UI with (webpack5), so that accounts for the extra install time. But build time is faster (20s vs ~30s before) with no optimizations added yet.

Storybook_Benchmark_›_Explorer__Install__Start__Build_

For static storybooks, time to first render is slightly faster (300ms vs 370ms before):

Storybook_Benchmark_›_Explorer__Browse_

**Comment by shilman** [comment]
@ndelangen we still need to set up an e2e test for webpack5, but can do it in a separate PR.

@gaetanmaisse I've broken the yarn2 PNP test but am merging anyway because my head will explode if I look at this PR any longer. Apologies for continuing to subject you to this.

**Comment by phaistonian** [comment]
@shilman  Somehow running `yarn storybook` `builder-webpack4` is preferred, as opposed to the `builder-webpack5` one.

I guess the detection phase needs some extra care.

Ps: any way to _force_ use the webpack5 one?

**Comment by shilman** [comment]
@phaistonian  yes you need to set it in `.storybook/main.js`

```js
module.exports = {
  core: { builder: 'webpack5' }
}
```

See the full migration guide here:

https://gist.github.com/shilman/8856ea1786dcd247139b47b270912324

**Pull request merged**